### ISSUE TITLE
OOMs just writing files with k8s memory limits (upstream kernel bug)

### ISSUE THREAD

**Issue by comstud** [coreos/bugs#2398]
# Issue Report #

## Bug ##

legacy memory cgroup proceses are OOM-killed just writing files. This is very evident when using CoreOS as a kubernetes node... configuring memory limits on pods. My upstream bug report can be found here which details the issue in a little more detail:

https://bugzilla.kernel.org/show_bug.cgi?id=199297

The fix for this is already committed upstream and is contained in kernel version 4.16. I am reporting it here, however, due to the fact that this should affect almost anyone using CoreOS as a kubernetes node. I am not sure how far off a stable CoreOS with kernel version 4.16+ would be... so perhaps this fix should be backported to a stable coreos release?

This is the upstream commit that fixes the issue:

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/commit/?id=1c610d5f93c7

### Container Linux Version ###

All versions with linux kernel 4.11 through (but not including) version 4.16 are affected.

### Environment ###

What hardware/cloud provider/hypervisor is being used to run Container Linux?
Doesn't matter.

### Expected Behavior ###

No OOM kills writing files.

### Actual Behavior ###

process OOM killed writing files (or write() syscall returns ENOMEM).

### Reproduction Steps ###

```
#!/bin/sh

set -e

if [ `id -u` != 0 ]; then
    echo 'must run as root'
    exit 1
fi

cg_memory_path_prefix='/sys/fs/cgroup/memory'

# Make a new cgroup and remove on exit
cg=`mktemp -p ${cg_memory_path_prefix} -d`
echo "Created $cg"
trap "{ echo 0 > ${cg}/memory.force_empty; rmdir $cg; }" EXIT

# 256M
echo 268435456 > ${cg}/memory.limit_in_bytes

# Add our pid to the cgroup
echo $$ > ${cg}/tasks

# Create tmp file
tmpfile=`mktemp -p /tmp`
trap "rm $tmpfile" EXIT

# Write 4G
echo "Writing to $tmpfile"
dd if=/dev/zero of=$tmpfile bs=1048576 count=4096
# Below not reached. dd will be OOMkilled.
echo 'No OOM'
```

**Comment by bgilbert** [comment]
Thanks for the report.  That patch is included in kernels &ge; 4.14.31 and &ge; 4.15.14.  The latest Container Linux stable release, 1688.5.3, thus already has the fix, and the next releases on the alpha and beta channels will as well.

**Comment by comstud** [comment]
Ah, so it is! Perfect. Thanks.
